### PR TITLE
[flex] better margin support with Wrap

### PR DESCRIPTION
**Pull request by StephaneDelcroix**
### Bugs Fixed ###

/

### API Changes ###

/

### Behavioral Changes ###

/

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [x] Consolidate commits as makes sense

### CODE REVIEW

File: Xamarin.Forms.Core/FlexLayout.cs [hunk 1]
```diff
@@ -408,12 +408,12 @@ protected override SizeRequest OnMeasure(double widthConstraint, double heightCo
 			if (double.IsPositiveInfinity(widthConstraint)) {
 				widthConstraint = 0;
 				foreach (var item in _root)
-					widthConstraint = Math.Max(widthConstraint, item.Frame[0] + item.Frame[2]);
+					widthConstraint = Math.Max(widthConstraint, item.Frame[0] + item.Frame[2] + item.MarginRight);
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/FlexLayout.cs]: this fixes measuring with margins

File: Xamarin.Flex/Flex.cs [hunk 2]
```diff
@@ -557,7 +555,7 @@ static void layout_item(Item item, float width, float height)
 					float basis = child.Basis.Length;
 					if (child.Basis.IsRelative)
 						basis *= (layout.vertical ? height : width);
-					child.Frame[layout.frame_size_i] = basis;
+					child.Frame[layout.frame_size_i] = basis - child.MarginThickness(layout.vertical);
```
Comment (StephaneDelcroix) [inline Xamarin.Flex/Flex.cs]: this removes the margin when using `basis` from the main-axis dimension

File: Xamarin.Flex/Flex.cs [hunk 3]
```diff
@@ -573,8 +571,8 @@ static void layout_item(Item item, float width, float height)
 					}
 
 					float child_size2 = child.Frame[layout.frame_size2_i];
-					if (!float.IsNaN(child_size2) && child_size2 > layout.line_dim) {
-						layout.line_dim = child_size2;
+					if (!float.IsNaN(child_size2) && child_size2 + child.MarginThickness(!layout.vertical) > layout.line_dim) {
+						layout.line_dim = child_size2 + child.MarginThickness(!layout.vertical);
```
Comment (StephaneDelcroix) [inline Xamarin.Flex/Flex.cs]: this grows the cross-axis dimension of the lines when using margins, and wrap

File: Xamarin.Flex/Flex.cs (hunk 2 above)
Comment (rmarinho) [reply Xamarin.Flex/Flex.cs]: How will this map to the code ported from flex lib? shouldn't at least we add a note that we mad this change?

Comment (StephaneDelcroix) [reply Xamarin.Flex/Flex.cs]: the code never mapped

**Pull request merged**